### DISCUSSION TITLE
How implent as modular Avalonia application with UI modules in .net assemblies?

### DISCUSSION THREAD

**Discussion by M0nteCarl0** [AvaloniaUI/Avalonia#16243] [Q&A]
Dear friends! 
Can you say best way to make Avalonia UI application with UI modules in assemblies?

**Comment by thevortexcloud** [comment]
Can you clarify what you mean by "UI modules"? Do you mean assemblies that contain user controls?

**Reply by M0nteCarl0** [reply]
Assemblies contains windows/dialogs with part of logic and ability dynamic load/unload.
Plugin system for my base CLI I am implemented via reflection and my custom IPlugin interface and class for resolving assemblies in specific folder. 
Now I want implement same behavior for my Avalonia application but with windows/dialogs in assemblies.

**Reply by thevortexcloud** [reply]
The process is the same in Avalonia. The only real difference is how you decide to load it into the UI. The simplest approach is to not load custom controls at all and instead just use standardised view models.

But if you do decide to load an actual control: I _think_ you may run into problems if you need to load custom style sheets as well. There are work arounds such as loading things at runtime you can do though.

https://docs.avaloniaui.net/docs/guides/styles-and-resources/how-to-use-included-styles

The typical approach with things like themes that add custom controls is to directly reference the style sheet in the App.axaml, which wont be possible if you don't know ahead of time what to load.

**Reply by M0nteCarl0** [reply]
I am first try implement Avalonia pluginable application without thems and custom controls. 
Am I right understand you is MVVM/Reactive is more preferred for implentation?

**Reply by thevortexcloud** [reply]
It depends on what you are doing. If you can get away with only using view models that are used by controls defined in your core app, then that will fine. If you need more advanced things than that, then it gets more complicated. I would expect most apps to do a bit of both approaches for this sort of thing.

**Reply by thevortexcloud** [reply]
I should note: If your external controls don't need custom style sheets, they should work with what ever the main app's one is.
